neuralop/layers/fno_block.py

```diff
@@ -190,74 +192,88 @@ def set_ada_in_embeddings(self, *embeddings):
-                    x_skip_mlp,
-                    self.output_scaling_factor[index],
-                    list(range(-len(self.output_scaling_factor[index]), 0)),
-                    output_shape=output_shape,
-                )
+            x_skip_mlp = self.resample(x_skip_mlp, index, output_shape)
 
         if self.stabilizer == "tanh":
             x = torch.tanh(x)
 
         x_fno = self.convs(x, index, output_shape=output_shape)
 
-        if not self.preactivation and self.norm is not None:
+        if self.norm is not None:
             x_fno = self.norm[self.n_norms * index](x_fno)
 
         x = x_fno + x_skip_fno
 
-        if (
-            not self.preactivation
-            and (self.mlp is not None)
-            or (index < (self.n_layers - index))
-        ):
+        if (self.mlp is not None) or (index < (self.n_layers - 1)):
             x = self.non_linearity(x)
 
         if self.mlp is not None:
-            # x_skip = self.mlp_skips[index](x)
+            x = self.mlp[index](x) + x_skip_mlp
 
-            if self.preactivation:
-                if index < (self.n_layers - 1):
-                    x = self.non_linearity(x)
+            if self.norm is not None:
+                x = self.norm[self.n_norms * index + 1](x)
 
-                if self.norm is not None:
-                    x = self.norm[self.n_norms * index + 1](x)
+            if index < (self.n_layers - 1):
+                x = self.non_linearity(x)
 
-            x = self.mlp[index](x) + x_skip_mlp
+        return x
+
+    def forward_with_preactivation(self, x, index=0, output_shape=None):
+        # Apply non-linear activation (and norm)
+        # before this block's convolution/forward pass:
+        x = self.non_linearity(x)
+
+        if self.norm is not None:
+            x = self.norm[self.n_norms * index](x)
 
-            if not self.preactivation and self.norm is not None:
+        x_skip_fno = self.fno_skips[index](x)
+        x_skip_fno = self.resample(x_skip_fno, index, output_shape)
+
+        if self.mlp is not None:
+            x_skip_mlp = self.mlp_skips[index](x)
+            x_skip_mlp = self.resample(x_skip_mlp, index, output_shape)
+
+        if self.stabilizer == "tanh":
+            x = torch.tanh(x)
+
+        x_fno = self.convs(x, index, output_shape=output_shape)
+        x = x_fno + x_skip_fno
+
+        if self.mlp is not None:
+            if index < (self.n_layers - 1):
+                x = self.non_linearity(x)
+
+            if self.norm is not None:
                 x = self.norm[self.n_norms * index + 1](x)
 
-            if not self.preactivation:
-                if index < (self.n_layers - 1):
-                    x = self.non_linearity(x)
+            x = self.mlp[index](x) + x_skip_mlp
+
         return x
 
+    def resample(self, x, index, output_shape):
+        """Resamples input if scaling factors are available for this block."""
+        if self.output_scaling_factor is None:
```

Comment:
Shouldn't you check both self.output_scaling_factor and output_shape vs x.shape?

Reply:
Good point. I've updated `FNOBlock.resample()` to defer first to arg `output_shape` to be passed to `resample`. AFAICT, when `output_shape is not None` in `resample.py:resample()` the other args are ignored (I passed dummy args `res_scale=1, axis=None`, so it should be fine).